### ISSUE TITLE
Hierarchical folder Including/Excluding/Hiding

### ISSUE THREAD

**Issue by UplinkPhobia** [SimpleMobileTools/Simple-Gallery#1027]
Hello,

An interesting feature I'd like to request, if possible, is a sort of priority between the three states (in/ex-cluded and hidden).
If I'm not mistaken:
- If a folder is both included and excluded, it will show.
- If a folder is excluded, but inside and included folder, it will show.
- If a folder is included, but inside an excluded folder, it will show.
- If a folder is included and hidden, it will not show.
- If a folder is hidden, but not its subfolders, they will show.
- Hiding a folder creates .nomedia files in all subfolders to "fix" the previous point.
- Unhiding a folder deletes **all** .nomedia files, including the ones that the app didn't create.

The way I see it, it's hard to grasp these priorities as they seem arbitrary and confusing.
That's why I'd suggest to take the path into account too. More explicitely, if a folder has a defined "state", it will apply to all subfolders **unless** they themselves have a defined state.

The behaviour would change to:
- If an included folder is inside an excluded/hidden one, it will still show.
- If an excluded/hidden folder is in a included one, it will not show.
- Hiding a folder only creates one .nomedia file, as the subfolders become implicitely hidden (unless stated otherwise)

Maybe I'm mistaken, but I feel that this would simplify the app's usage and allow for easier customisation of shown folders, while still removing issues with the .nomedia files.

Thanks in advance.

**Comment by UplinkPhobia** [comment]
Indeed, I got confused by a few things. It's my mistake.
However, subfolders of hidden folders seem to show on my phone anyway. Here's what I have:
- Excluded the root folder
- Included the "Pictures" folder
- Inside the "Pictures" folder, I have a "test" folder, with a picture
- Inside the "test" folder, I have a "testsub" folder, with a picture too

The two folders show up as expected, and I hide the first one ("test").
This makes the test folder disappear, but the "testsub" folder persists, with its image still shown.

(I'm using the last FDroid version (5.1.1), if it matters)

**Comment by SebiderSushi** [comment]
I'm also facing the issue in regards to "Include" taking absolute precedence over "Exclude", judging by its behaviour, and think it should take hierarchical precedence.

My case:
- I have set to include the folder `/storage/emulated/0/Pictures`
- The contents of that folder show up as expected on the next rescan.
- I then set to exclude the folder `/storage/emulated/0/Pictures/memetastic`. Now i expect all images within the `Pictures` folder to show up, except the ones from `memetastic`.
- I navigate back to the main screen and pull to refresh. The folder `memetastic` that i expect to be excluded now is still shown. From my perspective as the user it seems as though my explicit wish to exclude this folder is being ignored.

**Comment by tibbi** [comment]
@SebiderSushi  what app version are you using?

**Comment by SebiderSushi** [comment]
I am using Simple Gallery Pro version 6.9.4 from F-Droid.
Today when trying to reproduce my issue with gallery-release.apk version 6.9.5 from GitHub i noticed that maybe everything already works as it's supposed to.

What i observed today was that when adding a folder to the list of folders to exclude Simple Gallery would exclude that folder along with its children. If however, that folder would be located somewhere within another folder that has been added to the list of folders to include, then Simple Gallery would only exclude the folder specifically named in the exclusions list but still include any subfolders. I didn't notice that because up until today i never tried to hide one specific folder but rather just collections of folders. So i apologize for stating that the folder memetastic would not be excluded when i was referring to its subfolders that wouldn't disappear.

Now that i understand that this is likely not a bug and that there are use cases for the current behavior, let me rephrase my issue:

When adding a folder to the list of exclusions or inclusions i always expected Simple Gallery to consistently apply these restrictions recursively - as it already does, except when trying to exclude a folder while one of its parent folders is on the inclusions list.

What's confusing about this: I just set `/storage/emulated/0` to be excluded and then `/storage/emulated/0/Pictures` to be included. Now Simple Gallery recursively shows all folders within Pictures. It is not recursive when doing it the other way around, i.e. including `/storage/emulated/0` and exlcuding `/storage/emulated/0/Pictures`.

My normal use case requires the exclusions list to behave recursively even if a higher level directory is included. Its the only way i can, for example, include the folder Pictures and exclude memetastic without manually adding every single image folder within memetastic to the list of exclusions.

Since there are still use cases for the current behavior, i'd suggest a dialog when adding a folder to the exclusions or the inclusions list to manually specify whether Simple Gallery should act recursively, i.e. something among the lines of "Exclude only this folder" vs. "Exclude this folder and its subfolders".


I would be very happy if Simple Gallery could support my use case and i'd like to thank you for all of your time that you've put into this amazing app.

**Comment by tibbi** [comment]
It works as expected now and it is sufficient for 99% of usecases. You can get any functionality you want with some more work. But the vast majority of people dont use such complicated scenarios.

**Comment by SebiderSushi** [comment]
About "as expected" (from a users perspective): When selecting a folder on the main screen and choosing "Exclude" from the context menu the dialog window reads "This will exclude the selection together with its subfolders". This does not hold true in my scenario where i only have a higher level directory manually included and comes out to me as counterintuitive at least.

Any scenario currently supported by this behavior actually has to be "complicated" i.e. excluding something below an included folder. After all, this quirk _only ever shows_ in such a scenario. The only difference is that i expect something else than what happens.

Of course i'm not familiar with real world usage and you defininitely should now, dealing with all the other issues here on GitHub and if most users actually want to include `/foo`, exlcude `/foo/bar` but keep `/foo/bar/hello` then fine.

It just seems very weird to me because, to sum it up, what you're saying comes to me as: A counterintuitive behavior in a special case that is very hard to figure out as a cause to the behavior is more likely to be wanted than a consistent behavior as described even by the app itself.

If i'm exaggerating here just disregard the last paragraph, by now i'm merely trying to explain how weird it feels to me. In the end, you are the developer and you are the one working on the project so its only right for you to have the last word on any decision.

If it is your last word that you won't change anything about the current behavior: Would you accept pull requests that implement both usecases (always exclude subfolders plus the current behavior) or would this make the app unneccesarily complicated and i should be using a personal build instead?

**Comment by tibbi** [comment]
I dont want to overcomplicate it. As I said, it works just fine to 99% of users. People usually dont include and exclude the same folder..

**Comment by SebiderSushi** [comment]
But if no one usually includes `/foo` to exclude `/foo/bar` shouldn't it be fine to just change the behaviour to also exclude subfolders of `/foo/bar` like `/foo/bar/hello` without adding an option?

**Comment by tibbi** [comment]
some people expect one behaviour, some people the other one

**Comment by UplinkPhobia** [comment]
Indeed. And from the issues, it seems most people want this behaviour to be as we described.
Moreover, that's the normal behaviour of a .nomedia file is also the one we described. I don't see any valid reason to keep the current behaviour, and even less to refuse pull requests that would fix it.

It is a bug, not a feature. Not fixing it is at best understandable, refusing to have it fixed is plain irrational stubbornness.

**Comment by puppykickr** [comment]
It seems as though this issue truely stems from the fact that Memetastic and some others do not allow user control over where files are stored.

I have just entered a feature request for this on Memetastic.

Hopefully, if implemented, the user can then change the folder used by Memetastic for saving to another one that can be hidden or not as to the user's liking.

**Comment by SebiderSushi** [comment]
@puppykickr Thank You. Being able to move the Memetastic folder would, for me, solve the concrete issue that i'm currently trying to solve since as long as i'm able to move a folder i can always work around the strange exclude behavior from Simple Gallery by just doing that.


Nonetheless i'm still curious about the use cases of the current non-recursive-if-within-included-directory exclude behavior.

**Comment by puppykickr** [comment]
Here is my issue with your issue (Does that sentence make sense?  Oh well.).

I don't understand the need to include the 'Pictures' folder.

That is because on my device, all of the subfolders inside 'Pictures' appear without any instigation on my part.

This would make it very simple to hide 'Memetastic' and then it would be hidden but the other subfolders would show up as usual.

Pictures itself has no need of being manually included-  at least on my device-  which is nothing special and far from being expensive.

Something else I have done is make a 'My Memes' folder on my SD card, where I move my finished memes after creation.

This folder shows up on its own in SimpleGallery, without any action on my part.

That being said, if I were to hide that folder, then it would not appear.

Until Memetastic allows for the user to choose the save location, I think this is your most obvious solution.

**Comment by SebiderSushi** [comment]
@puppykickr I for my part want my Gallery to be simple and cleaned up. That is why i prefer it to only show images from explicitly specified locations. For this reason i add `/` to the list of exclusions in Simple Gallery and include everything i actually want.

**Comment by puppykickr** [comment]
Well, the dev at Memetastic responded, and it appears that he thought he had already included the feature (selectable storage location) but had not.

So hopefully this will be included in a future update of Memetastic.

And as for your issue with Simple Gallery, and don't forget that I had the same issue as you do about a year or so ago, the simple solution turned out to be the correct one.

Allow the app to include whatever it desires naturally from the beginning as default.

Then hide whatever you don't want to see all the time.

Now, for me, after that I went back a d included some folders that get clogged with images, ads, shared items, etc.

I then hid them.

That way, I can easily 'Temporarily Show Hidden' and use SimpleGallery to find these files and delete them.

I found (and I am sure you will too) that working with the app is much more productive than working against it.

Anything you don't want to see in the gallery, make a separate folder for it then just hide that folder.
